core/vdbe/sorter.rs

```diff
@@ -32,26 +70,290 @@ impl Sorter {
+            self.flush()?;
+        }
+        self.records.push(SortableImmutableRecord::new(
+            record.clone(),
+            self.key_len,
+            self.order,
+            self.collations.clone(),
+        ));
+        self.current_buffer_size += payload_size;
+        self.max_payload_size_in_buffer = self.max_payload_size_in_buffer.max(payload_size);
+        self.max_values_len_in_buffer = self.max_values_len_in_buffer.max(record.len());
+        Ok(())
+    }
+
+    fn push_to_heap_from_chunk(&mut self, chunk_idx: usize, init: bool) -> Result<()> {
+        let chunk_reader = &mut self.chunk_readers[chunk_idx];
+
+        if init {
+            chunk_reader.next()?;
+        }
+
+        if chunk_reader.has_more() {
+            let record = chunk_reader.consume_record()?.unwrap();
+            self.chunk_heap.push((
+                Reverse(SortableImmutableRecord::new(
+                    record,
+                    self.key_len,
+                    self.order,
+                    self.collations.clone(),
+                )),
+                chunk_idx,
+            ));
+        }
+        Ok(())
+    }
+
+    fn flush(&mut self) -> Result<()> {
+        self.records.sort();
+
+        let mut fd = tempfile::tempfile()?;
+        let mut writer = io::BufWriter::new(fd.try_clone()?);
+        for record in self.records.drain(..) {
+            writer
+                .write(record.record.get_payload())
+                .map_err(LimboError::IOError)?;
+        }
+        writer.flush().map_err(LimboError::IOError)?;
+
+        fd.seek(io::SeekFrom::Start(0))
+            .map_err(LimboError::IOError)?;
+        self.chunk_readers.push(ChunkReader::new(
+            fd,
+            self.max_payload_size_in_buffer,
+            self.max_values_len_in_buffer,
+        ));
+
+        self.current_buffer_size = 0;
+        self.max_payload_size_in_buffer = 0;
+        self.max_values_len_in_buffer = 0;
+
+        Ok(())
+    }
+}
+
+struct ChunkReader {
+    /// The reader for the chunk file.
+    reader: io::BufReader<File>,
+    /// The read buffer.
+    buffer: Vec<u8>,
+    /// The current length of the buffer.
+    buffer_len: usize,
+    /// The maximum number of values in a record in this chunk.
+    max_values_len: usize,
+    /// The current record.
+    current: Option<ImmutableRecord>,
+}
+
+impl ChunkReader {
+    fn new(fd: File, max_payload_size: usize, max_values_len: usize) -> Self {
+        Self {
+            reader: io::BufReader::new(fd),
+            buffer: vec![0; max_payload_size],
+            buffer_len: 0, // Start with empty buffer
+            max_values_len,
+            current: None,
+        }
+    }
+
+    fn has_more(&self) -> bool {
+        self.current.is_some()
+    }
+
+    fn consume_record(&mut self) -> Result<Option<ImmutableRecord>> {
+        let result = self.current.take();
+        self.next()?;
+        Ok(result)
+    }
+
+    fn next(&mut self) -> Result<()> {
```

Comment:
Out of curiosity, is there a reason you chose not to implement the `Iterator` trait for `ChunkReader`? I've noticed that implementing the `Iterator` pattern gives you nicer ergonomics.

Reply:
Yeah, I ran into similar issues in the past. I went hunting around and found an implementation of [fallible iterators]((https://docs.rs/fallible-iterator/latest/fallible_iterator/)).

It's unmaintained and probably not worth introducing into Limbo but just in case you haven't seen it or are interested.